Script Updates

jason-oliveira opened the pull request:
* Now works on non-Gentoo LiveCD's.
* GENTOO_STAGE3 profile changed to "amd64-openrc"
* TARGET_DISK is now /dev/vda, because it should be.
* /boot is 256M now.
* Using Calculate Linux as optional binhost for faster install.
* Genkernel is no longer needed - installing binary kernel packages.
* make.conf should be a directory.
* Licenses needed to be accepted.
* dhcpcd added so network doesn't fail on reboot.
* ~~eix added/synced~~

Commit:
Updates were necessary.

Commit:
Added eix, dhcpcd, and more mount options for Non-Gentoo LiveCD's.

Commit:
Reorganized lines after I fixed the Binhost line.

sormy commented on gentoo-quick-installer.sh:
```diff
@@ -16,18 +16,24 @@
 
 GENTOO_ARCH="amd64"
-GENTOO_STAGE3="amd64"
+GENTOO_STAGE3="	amd64-hardened-nomultilib-openrc"
```
lets keep amd64 as default one, people can update it if needed manually

sormy commented on gentoo-quick-installer.sh:
```diff
@@ -16,18 +16,24 @@
+GENTOO_STAGE3="	amd64-hardened-nomultilib-openrc"
 
-TARGET_DISK=/dev/sda
+TARGET_DISK=/dev/vda
```
Same, disk naming convention purely depends on target VM.

sormy commented on gentoo-quick-installer.sh:
```diff
@@ -16,18 +16,24 @@
+TARGET_DISK=/dev/vda
 
-TARGET_BOOT_SIZE=100M
+TARGET_BOOT_SIZE=256M
```
Isn't 100M enough?

sormy commented:
Looks like a completely different script.

I tested it a long time ago but what are absolutely necessary changes and what can be omitted?

Changing default boot size, disk name, pulling mandatory yandex binhost that could have availability problems due to sanctions, installing eix, using genkernel when it is not used to build the kernel, adding `-g` to use binary, adding shm and tmps that are supposed to be here already, --rbind - is it all required to make it work? 

Pulling binary kernel is actually a good thing, LIVE_CD code path can be removed if binary kernel is working well and allow to build kernel extensions etc. Or live cd code path can be replaced with binary kernel. The whole purpose of LIVE CD KERNEL code path is to make install faster.

Using binary host is probably a good idea but we might need to pull it in separate script because it impacts code too much. Also, default server for packages should not point to yandex mirror if we wanted to make it work worldwide for sure. I would prefer to keep it more vanilla without using binhost but it is ok to use binary packages from gentoo repo.

jason-oliveira commented:
Thanks for your responses.

* 100MB holds exactly one kernel in /boot. du -sh shows 54MB on the dir after install. [Even the Gentoo Handbook recommends 256MB for /boot](https://wiki.gentoo.org/wiki/Handbook:AMD64/Installation/Disks#Default_partitioning_scheme).
* ~~Calculate Linux is a russian distribution, and the best source of Stage4 binaries. I would've chose CloverOS or Sabayon if the packages were current or closer to upstream Gentoo. The URL can be (should be) replaced with a generic roundrobin mirror.~~
* ~~`emerge -g` doesn't force binaries if they're not available. comment-out the binhost line, and everything installs from source without further modification. (TODO: comment this in script)~~
* For many people (such as myself) running this script on a VPS inside a non-Gentoo LiveCD, the modifications to `shm` and `tmpfs` are essential to ensure chroots work.
* More people are installing Gentoo on remote server/VPS instances rather than bare metal, which is why /dev/vda was chosen over /dev/sda. A comment for /dev/sda and /dev/nvme0n1 could also be included.
* If this wasn't a server install script, then you wouldn't have included the ability to log in via pubkey only and delete root password (remote access only), so this is Reason Number 2 why /dev/vda is chosen.
* There is no "amd64", so it would be changed to "amd64-openrc" at least, which is ridiculously vulnerable for a server-oriented install script, post heartbleed/retbleed. ~~(TODO: comment other options available)~~
* Disk name was not changed. If it was, that was an oversight.
* If you're worried about --rbind, please [note the Gentoo Handbook](https://wiki.gentoo.org/wiki/Handbook:AMD64/Installation/Base#Mounting_the_necessary_filesystems). I omitted the systemd-required lines, if you'll notice, but they could be added with little issue. I just don't like systemd, so I don't go out of my way to support.
* ~~I'll make deleting the genkernel lines a TODO. I left them in just in case someone preferred manual compilation, and merely commented out the genkernel build line. More than happy to remove them.~~

More than happy to compromise. it's your project. :+1:

Commit:
removed genkernel, and commented changes with other options.

Commit:
...And he signs his name with a Capital G (by mistake)

Commit:
removed anything that could be considered superfluous, and added comments.

jason-oliveira commented:
* Removed eix, as it isn't necessary for base install.
* Removed genkernel, as it isn't needed, due to binary kernel.
* Added more documentation to top of script, including limitations (no UEFI support yet, GPT limitation)
* Removed all `emerge -g` and PORTAGE_BINHOST references.
* Removed ~amd64 kernel default. Stable is better.
* GENTOO_STAGE3 set to `amd64-openrc` by default.

Ideas that likely deserve separate issues/PR's outside the scope of this script update:
* ~~UEFI support - /boot is efi/vfat, partition table is GPT, and `grub` needs additional flags.~~ #3
* ~~Deprecating USE_LIVECD_KERNEL because Gentoo makes this redundant/irrelevant.~~ #8
* ~~Actual robust PORTAGE_BINHOST support. There likely needs to be an array set for binhost mirrors (should they need to be manually added - unable to find roundrobin), and if the variable is unset/null, then no binhost is written.~~ #9
* ~~if ROOT_PASSWORD is set, /dev/{sda/nvme0/mmcblk0) can be detected, and TARGET_DISK can be modified accordingly.~~ #7

jason-oliveira committed:
Updating README.md for clarity and limitations.

Commit:
Making the drive mounting compliant with Gentoo Handbook.

jason-oliveira commented:
Gentoo Handbook Changes to Mount:
* The `-o bind` behavior for `mount` is deprecated.
* `--rbind` allows a recursive bind on mounts, eliminating two `/dev` mounts.
* `/dev/shm` fixes given proper echoed heading for fixing broken LiveCD's.

Commit:
forgot to remove one -g.

jason-oliveira commented:
Alright, I think this as barebones as it can be now. Let me know if more changes are required.

Commit:
forgot to remove genkernel's emerge.

jason-oliveira commented:
@sormy let me know if anything needs to be changed. Once this PR is merged, I'l begin working on the other issues.

Commit:
How did that space get into GENTOO_STAGE3?

sormy commented on gentoo-quick-installer.sh:
```diff
@@ -189,6 +204,7 @@ echo "### Configuring network..."
 
 ln -s /etc/init.d/net.lo /etc/init.d/net.eth0
 rc-update add net.eth0 default
+emerge net-misc/dhcpcd
```
is not gentoo uses dhclient from busybox by default?

jason-oliveira replied:
I just tested for dhclient, and it was not installed. 

Funny enough, the Gentoo Handbook [mentions this specifically](https://wiki.gentoo.org/wiki/Handbook:AMD64/Installation/Tools#Installing_a_DHCP_client) with a big warning at the top telling you that if you reboot without installing, you won't have network connectivity. :laughing:

sormy replied:
Well, then lets keep it here

sormy commented on gentoo-quick-installer.sh:
```diff
@@ -148,27 +165,25 @@ echo "### Installing portage..."
+# required to allow for linux-firmware (required for binary kernel).
+echo "ACCEPT_LICENSE=\"*\"" >> /etc/portage/make.conf/default
 
 if [ "$USE_LIVECD_KERNEL" = 0 ]; then
```
So I suggest to have something like GENTOO_BINARY=0/1. This flag will help to install in fastest way all binary. Otherwise standard kernel installation and package compilation will be used. What do you think? And USE_LIVECD_KERNEL can be deprecated with relevant code that copies kernel.

jason-oliveira replied:
This is a very interesting question. Is the purpose of the script to install packages as fast as possible? 

I would be inclined to say that the entire script should be focused on installing via binary, and that compilation should be avoided if at all possible. As [OSU's OSL's Stage 4](https://gentoo.osuosl.org/experimental/amd64/binpkg/default/linux/17.1/x86-64/) exists and is considered official, the only logical reason for source building is if you wanted to enable -march=native in the CFLAGS for that extra 1-2% performance.

If you really wanted to include a GENTOO_BINARY flag, probably best to just set it to 1 internally. Otherwise, I'd just omit the variable entirely, and set it. This issue might be out of the scope of this update, and I'll do a separate PR for binary and `emerge -g` to keep this one simple to merge. We can absolutely remove USE_LIVECD_KERNEL, though.

sormy replied:
Building kernel from the source can be needed if you want to build custom kernel module.
This CR removes completely this option, i would keep it, the rest is good.

sormy commented on gentoo-quick-installer.sh:
```diff
@@ -11,23 +11,41 @@
+# /dev/sda is standard for most IDE/SATA drives.
+# /dev/nvme0n1 is standard for NVME drives.
+# /dev/mmcblk0 is standard for most eMMC and SD drives
+TARGET_DISK=/dev/vda
```
Might make it `TARGET_DISK=${TARGET_DISK:-/dev/sda}`.
This script was created for installation in VirtualBox. We can keep default for backward compatibility but still allowing to override disk name.

jason-oliveira replied:
The issue with this idea is the weird hybrid it creates with the script, where you're either expected to hand-edit it (GENTOO_STAGE3 and TARGET_DISK) to make sure it works, and issue envars while running the command afterwards. for the purposes of automation and being run from other scripts, hand-editing scripts is less desireable. 

I'll change both TARGET_DISK and GENTOO_STAGE3 for now.

sormy replied:
Agree

Commit:
Requested changes made
